src/core/desktopentry.cpp

```diff
@@ -407,7 +440,115 @@ DesktopEntry* DesktopEntryManager::heuristicLookup(const QString& name) {
 
 ObjectModel<DesktopEntry>* DesktopEntryManager::applications() { return &this->mApplications; }
 
-DesktopEntries::DesktopEntries() { DesktopEntryManager::instance(); }
+void DesktopEntryManager::handleFileChanges() {
+	qCDebug(logDesktopEntry) << "Directory change detected, performing full rescan";
+
+	if (this->scanInProgress) {
+		qCDebug(logDesktopEntry) << "Scan already in progress, skipping";
+		return;
+	}
```

Comment:
This looks race-prone. Consider the following case:
1. An app is added to a scan dir, we begin a scan
2. Another app is added once we've already ls'd the dir, scanInProgress is set so it does nothing
3. We finish the scan, unset scanInProgress, and don't find the other app

I would recommend a scanQueued flag or equivalent that kicks off another scan when we'd set scanInProgress to false, ref desktopentry.cpp:496